### PR TITLE
Improve Search Autocomplete UI

### PR DESCRIPTION
**Pull request by meowsus**

### CODE REVIEW

File: test/system/workarea/storefront/search_autocomplete_system_test.rb [hunk 1]
```diff
@@ -18,9 +18,16 @@ def test_autocomplete
...
+        fill_in 'q', with: ''
 
         assert_selector('#search_autocomplete', visible: false)
+
+        fill_in 'q', with: 'foo'
+
+        within '#search_autocomplete' do
+          assert_text(t('workarea.storefront.search_autocomplete.trending_products'))
+          assert_text(t('workarea.storefront.search_autocomplete.trending_searches'))
+        end
```
Comment (meowsus) [inline test/system/workarea/storefront/search_autocomplete_system_test.rb]: Looking for input on a good way to generate and test Trending Product/Search results. Also for creating a search customization and content to prove that content can appear in the autocomplete.

File: app/view_models/workarea/storefront/search_autocomplete_product_view_model.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
+module Workarea
+  module Storefront
+    class SearchAutocompleteProductViewModel < RecommendationsViewModel
```
Comment (bencrouse) [inline app/view_models/workarea/storefront/search_autocomplete_product_view_model.rb]: Since this doesn't wrap a single product (it presents a collection of them), I think it'd make more sense to reflect this in the class name. Like `SearchAutocompleteProductsViewModel` or `SearchAutocompleteRecommendationsViewModel` or just something plural.

File: app/view_models/workarea/storefront/search_autocomplete_view_model.rb [hunk 3]
```diff
@@ -3,6 +3,15 @@ module Storefront
     class SearchAutocompleteViewModel < ApplicationViewModel
       def products
         return [] if searches.blank?
+        return product_results unless results.blank?
+
+        SearchAutocompleteProductViewModel.wrap(
```
Comment (bencrouse) [inline app/view_models/workarea/storefront/search_autocomplete_view_model.rb]: We'll want to memoize this since it's a fairly expensive thing to grab.

File: app/views/workarea/storefront/searches/autocomplete.html.haml [hunk 4]
```diff
@@ -1,15 +1,29 @@
 .grid.grid--rev
   .grid__cell.grid__cell--25
     %ul.search-autocomplete__searches
+      - if @autocomplete.results.blank?
+        %span.search-autocomplete__heading= t('workarea.storefront.search_autocomplete.trending_searches')
```
Comment (bencrouse) [inline app/views/workarea/storefront/searches/autocomplete.html.haml]: Is it kosher to put a `span` under a `ul`?

File: config/initializers/content.rb [hunk 5]
```diff
@@ -0,0 +1,5 @@
+Workarea.configure do |config|
+  config.content_areas.merge!(
```
Comment (bencrouse) [inline config/initializers/content.rb]: (minor) A simpler way to do this would be `config.content_areas['customization'].append('autocomplete')`

File: app/controllers/workarea/storefront/searches_controller.decorator [hunk 6]
```diff
@@ -6,7 +6,7 @@ module Workarea
         view_model_options
       )
 
-      head 404 if @autocomplete.no_results?
+      @search = Storefront::SearchViewModel.new(@autocomplete.response, view_model_options)
```
Comment (bencrouse) [inline app/controllers/workarea/storefront/searches_controller.decorator]: Can we add content to the autocomplete view model instead? Easier to test, keeps the number of view models at a minimum, less confusing in the view.

File: app/view_models/workarea/storefront/search_autocomplete_view_model.rb [hunk 7]
```diff
@@ -15,15 +24,19 @@ def response
       end
 
       def searches
-        @searches ||= Metrics::SearchByWeek.autocomplete(
+        return results unless results.blank?
+
+        Workarea::Insights::TrendingSearches.current.results.map do |v|
```
Comment (bencrouse) [inline app/view_models/workarea/storefront/search_autocomplete_view_model.rb]: This should also be memoized, since otherwise it will do the query more than once.

File: app/view_models/workarea/storefront/search_autocomplete_view_model.rb [hunk 8]
```diff
@@ -15,15 +24,19 @@ def response
...
+          v['query_id']
+        end
+      end
+
+      def results
```
Comment (bencrouse) [inline app/view_models/workarea/storefront/search_autocomplete_view_model.rb]: Why switch this with `searches`? It would read nicer in the view if it were something like
```
  - if @autocomplete.searches.blank?
    - @autocomplete.trending_searches.each do |search|
```

Which the language then describes the functionality which is to fall back to trending searches if the autocompleted searches are empty.

File: test/system/workarea/storefront/search_autocomplete_system_test.rb (hunk 1 above)
Comment (bencrouse) [reply test/system/workarea/storefront/search_autocomplete_system_test.rb]: For the trending product/searches, try checking out the unit tests for those classes.
For the search autocomplete, I believe there's a storefront system test for that.

File: app/view_models/workarea/storefront/search_autocomplete_view_model.rb [hunk 9]
```diff
@@ -2,10 +2,16 @@ module Workarea
   module Storefront
     class SearchAutocompleteViewModel < ApplicationViewModel
       def products
-        return [] if searches.blank?
```
Comment (bencrouse) [inline app/view_models/workarea/storefront/search_autocomplete_view_model.rb]: I think we'd still want this? This was here to short-circuit running the whole query (pretty expensive) if we know that there aren't any searches to match on.

File: app/view_models/workarea/storefront/search_autocomplete_view_model.rb [hunk 10]
```diff
@@ -21,8 +27,16 @@ def searches
         )
       end
 
-      def no_results?
-        products.empty? && searches.empty?
+      def trending_searches
+        @trending_searches ||= begin
+          Workarea::Insights::TrendingSearches.current.results.map do |v|
+            v['query_id']
```
Comment (bencrouse) [inline app/view_models/workarea/storefront/search_autocomplete_view_model.rb]: The `query_id` helps group similar queries (e.g. `shirts` and `shirt`), so we'll want to use the full query string to avoid output like `red_shirt`.

File: test/system/workarea/storefront/search_autocomplete_system_test.rb [hunk 11]
```diff
@@ -18,9 +144,18 @@ def test_autocomplete
           assert_match(/test two.*test one/m, page.body)
         end
 
-        fill_in 'q', with: 'foo'
+        fill_in 'q', with: ''
 
         assert_selector('#search_autocomplete', visible: false)
+
+        fill_in 'q', with: 'corge'
```
Comment (tubbo) [inline test/system/workarea/storefront/search_autocomplete_system_test.rb]: you spelled "doge" wrong.

File: app/view_models/workarea/storefront/search_autocomplete_view_model.rb (hunk 10 above)
Comment (bencrouse) [reply app/view_models/workarea/storefront/search_autocomplete_view_model.rb]: Doesn't seem like this is resolved? You'll want to output the `query_string`.

File: app/assets/stylesheets/workarea/storefront/search_autocomplete/components/_search_autocomplete.scss [hunk 12]
```diff
@@ -39,14 +38,14 @@ $search-autocomplete-z-indexes: ();
 }
 
 
+    .search-autocomplete__content,
```
Comment (tubbo) [inline app/assets/stylesheets/workarea/storefront/search_autocomplete/components/_search_autocomplete.scss]: this should probably be in content-search...

**Comment by tubbo** [comment]
I think you need to wrap the route in a `scope` block, otherwise you can't pass a locale through to it. here's what the route looks like when you run `rails app:routes`:

```
autocomplete_search GET    /search/autocomplete(.:format)                            workarea/storefront/searches#autocomplete
```

for comparison, here's the route coming out of base:

```
search GET    (/:locale)/search(.:format)                               workarea/storefront/searches#show {:locale=>/en/}
```

**Comment by bencrouse** [comment]
@tubbo that is a good, but unrelated catch. can you open a separate PR?

File: app/view_models/workarea/storefront/search_autocomplete_view_model.rb (hunk 10 above)
Comment (bencrouse) [reply app/view_models/workarea/storefront/search_autocomplete_view_model.rb]: @meowsus looks like this is still outputting `query_id`

Comment (meowsus) [reply app/view_models/workarea/storefront/search_autocomplete_view_model.rb]: Derf. Missed this.

File: test/system/workarea/storefront/search_autocomplete_system_test.rb (hunk 11 above)
Comment (meowsus) [reply test/system/workarea/storefront/search_autocomplete_system_test.rb]: http://catb.org/jargon/html/

File: test/system/workarea/storefront/search_autocomplete_system_test.rb [hunk 13]
```diff
@@ -18,9 +144,19 @@ def test_autocomplete
...
+
+        within '#search_autocomplete' do
+          assert_text(t('workarea.storefront.search_autocomplete.trending_products'))
+          assert_text(t('workarea.storefront.search_autocomplete.trending_searches'))
+          save_and_open_screenshot
```
Comment (tubbo) [inline test/system/workarea/storefront/search_autocomplete_system_test.rb]: probably don't need this anymore right?

**Comment by meowsus** [comment]
This PR uncovered a bug in base that @bencrouse is fixing currently. Once it's merged I'll hit the build again.

File: app/view_models/workarea/storefront/search_autocomplete_view_model.rb [hunk 14]
```diff
@@ -21,8 +28,16 @@ def searches
...
-        products.empty? && searches.empty?
+      def trending_searches
+        @trending_searches ||= begin
+          Workarea::Insights::TrendingSearches.current.results.map do |v|
+            v['query_string']
+          end
+        end
+      end
+
+      def content
```
Comment (tubbo) [inline app/view_models/workarea/storefront/search_autocomplete_view_model.rb]: i defined this in ContentSearch so you won't need it here.

Comment (bencrouse) [reply app/view_models/workarea/storefront/search_autocomplete_view_model.rb]: This is for outputting content blocks (see the autocomplete view). We'll need both.

Comment (tubbo) [reply app/view_models/workarea/storefront/search_autocomplete_view_model.rb]: so should the method I defined be called `#contents` or something?
